### DISCUSSION TITLE
Mismatched IDs received by the Tok2Vec listener while training

### DISCUSSION THREAD

**Discussion by DSLituiev** [explosion/spaCy#7387] [Help: Coding & Implementations] [answered]
I'm trying to run training from python and running into an issue

```python

train_docs_ents = []
for text, annotations in tqdm(train_data):
    train_docs_ents.append(Example.from_dict(nlp.make_doc(text),  annotations))
   # where annotations are of the form:  (start, end, label)

n_iter = 80
        
# batch_size = compounding(4.0, 32.0, 1.001)
batch_size = 16
with nlp.select_pipes(enable="ner"):  # only train NER
    optimizer = nlp.initialize()
#     optimizer = None
    for itn in (range(n_iter)):
        random.shuffle(fixed_train_data)
        losses = {}
        
        batches = minibatch(train_docs_ents, size=batch_size)
        for batch in tqdm(batches, total= math.ceil(len(train_docs_ents)/batch_size)):
            print("1")
            nlp.update(batch,  
                drop=0.5,  
                sgd=optimizer,
                losses=losses)
        logging.info(f'loss: {losses["ner"]}')
```
I'm running into an error:

```
-------------------------------------------------------------------
ValueError                        Traceback (most recent call last)
<ipython-input-50-5e53522a10c2> in <module>
     13         for batch in tqdm(batches, total= math.ceil(len(fixed_train_docs_ents)/batch_size)):
     14             print("1")
---> 15             nlp.update(batch,  
     16                 drop=0.5,
     17                 sgd=optimizer,

/data/lituievd/anaconda3/envs/pathnlp/lib/python3.8/site-packages/spacy/language.py in update(self, examples, _, drop, sgd, losses, component_cfg, exclude)
   1104         for name, proc in self.pipeline:
   1105             if name in exclude or not hasattr(proc, "update"):
-> 1106                 continue
   1107             proc.update(examples, sgd=None, losses=losses, **component_cfg[name])
   1108         if sgd not in (None, False):

/data/lituievd/anaconda3/envs/pathnlp/lib/python3.8/site-packages/spacy/pipeline/transition_parser.pyx in spacy.pipeline.transition_parser.Parser.update()

/data/lituievd/anaconda3/envs/pathnlp/lib/python3.8/site-packages/thinc/model.py in begin_update(self, X)
    304         and returns the gradient with respect to the input.
    305         """
--> 306         return self._func(self, X, is_train=True)
    307 
    308     def predict(self, X: InT) -> OutT:

/data/lituievd/anaconda3/envs/pathnlp/lib/python3.8/site-packages/spacy/ml/tb_framework.py in forward(model, X, is_train)
     29 
     30 def forward(model, X, is_train):
---> 31     step_model = ParserStepModel(
     32         X,
     33         model.layers,

/data/lituievd/anaconda3/envs/pathnlp/lib/python3.8/site-packages/spacy/ml/parser_model.pyx in spacy.ml.parser_model.ParserStepModel.__init__()

/data/lituievd/anaconda3/envs/pathnlp/lib/python3.8/site-packages/thinc/model.py in __call__(self, X, is_train)
    286         """Call the model's `forward` function, returning the output and a
    287         callback to compute the gradients via backpropagation."""
--> 288         return self._func(self, X, is_train=is_train)
    289 
    290     def initialize(self, X: Optional[InT] = None, Y: Optional[OutT] = None) -> "Model":

/data/lituievd/anaconda3/envs/pathnlp/lib/python3.8/site-packages/thinc/layers/chain.py in forward(model, X, is_train)
     52     callbacks = []
     53     for layer in model.layers:
---> 54         Y, inc_layer_grad = layer(X, is_train=is_train)
     55         callbacks.append(inc_layer_grad)
     56         X = Y

/data/lituievd/anaconda3/envs/pathnlp/lib/python3.8/site-packages/thinc/model.py in __call__(self, X, is_train)
    286         """Call the model's `forward` function, returning the output and a
    287         callback to compute the gradients via backpropagation."""
--> 288         return self._func(self, X, is_train=is_train)
    289 
    290     def initialize(self, X: Optional[InT] = None, Y: Optional[OutT] = None) -> "Model":

/data/lituievd/anaconda3/envs/pathnlp/lib/python3.8/site-packages/spacy/pipeline/tok2vec.py in forward(model, inputs, is_train)
    282     """Supply the outputs from the upstream Tok2Vec component."""
    283     if is_train:
--> 284         model.verify_inputs(inputs)
    285         return model._outputs, model._backprop
    286     else:

/data/lituievd/anaconda3/envs/pathnlp/lib/python3.8/site-packages/spacy/pipeline/tok2vec.py in verify_inputs(self, inputs)
    274             batch_id = self.get_batch_id(inputs)
    275             if batch_id != self._batch_id:
--> 276                 raise ValueError(Errors.E953.format(id1=batch_id, id2=self._batch_id))
    277             else:
    278                 return True

ValueError: [E953] Mismatched IDs received by the Tok2Vec listener: 189341563201196604002545 vs. 12655626504890682208767
```

Any idea what could have gone wrong?
Unfortunately, I cannot provide data (healthcare).

**Comment by svlandeg** [comment, marked as answer]
Your NER model uses a `Tok2VecListener` that refers back to the `tok2vec` component in your pipeline (I assume).
When the NER tries to make a prediction, it will first fetch the token vectors from this `tok2vec` component, and when doing so, will include a check that it's getting the right token vectors for the right batch. This check fails in this case, because the `tok2vec` component has been disabled:
```
with nlp.select_pipes(enable="ner")
```

TLDR: you'll need to change this to 
```
with nlp.select_pipes(enable=["tok2vec","ner"])
```

FYI - we really recommend using the [config system](https://spacy.io/usage/training#config) for training your custom pipeline. We realise that there's a little bit of a learning curve, but once you start using it you'll see it's actually very powerful and customizable, and these kind of details will be taken care of for you - or at least a more useful error will be provided.
